test/assignment-5/CollateralizedVault.t.sol

```diff
@@ -0,0 +1,270 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+
+import "forge-std/Test.sol";
+import {CollateralizedVault} from "src/assignment-5/CollateralizedVault.sol";
+import {Dai} from "src/assignment-5/DAI.sol";
+import {WETH9} from "src/assignment-5/WETH.sol";
+import {IERC20} from "yield-utils-v2/token/IERC20.sol";
+import {ChainlinkPriceFeedMock} from "src/assignment-5/ChainlinkPriceFeedMock.sol";
+
+abstract contract ZeroState is Test {
+
+    using stdStorage for StdStorage;
+
+    address USER = address(1);
+    CollateralizedVault vault;
+    Dai dai;
+    WETH9 weth;
+    ChainlinkPriceFeedMock priceFeedMock;
+
+    function setUp() public virtual {
+        dai = new Dai(block.chainid);
+        weth = new WETH9();
+        priceFeedMock = new ChainlinkPriceFeedMock();
+        priceFeedMock.setPrice(500000000000000); // = 1/2000
+        vault = new CollateralizedVault(address(dai), address(weth), address(priceFeedMock));
+
+        setDaiBalance(address(vault), 10000 ether);
+        setWethBalance(USER, 10 ether);
+        vm.prank(USER);
+        weth.approve(address(vault), 10 ether);
+    }
+
+    function setDaiBalance(address dst, uint256 balance) public {
+        stdstore
+            .target(address(dai))
+            .sig(dai.balanceOf.selector)
+            .with_key(dst)
+            .depth(0)
+            .checked_write(balance);
+    }
+
+    function setWethBalance(address dst, uint256 balance) public {
+        stdstore
+            .target(address(weth))
+            .sig(weth.balanceOf.selector)
+            .with_key(dst)
+            .depth(0)
+            .checked_write(balance);
+    }
+}
+
+contract ZeroStateTest is ZeroState {
+
+    function testDeposit() public {
+        vm.prank(USER);
+        vault.deposit(3 ether);
+
+        // 3 WETH was transfered to the vault
+        assertEq(weth.balanceOf(USER), 7 ether);
+        assertEq(weth.balanceOf(address(vault)), 3 ether);
+
+        // User collateral is recorded
+        assertEq(vault.deposits(USER), 3 ether);
+    }
+
+    function testDepositMultipleTimesAccumulatesCollateral() public {
+        vm.startPrank(USER);
+
+        vault.deposit(1 ether);
+        vault.deposit(1 ether);
+
+        assertEq(weth.balanceOf(address(vault)), 2 ether);
+        assertEq(vault.deposits(USER), 2 ether);
+    }
+
+    function testScaleInteger() public {
+        // scale 12.345 from 6 decimals to 18
+        uint256 from = (12345 * 1e6) / 1e3;
+        uint256 to = vault.scaleInteger(from, 6, 18);
+
+        assertEq(to, (12345 * 1e18) / 1e3);
+
+        // test reverse
+        assertEq(vault.scaleInteger(to, 18, 6), from);
+    }
+}
+
+abstract contract DepositedCollateralState is ZeroState {
```

Comment:
i like the way you ordered this.  its very easy to read.  i had been grouping all my abstract contracts at the top but this is way better.  the only thing i would change is it would be nice if the test that proves u can do whatever u r doing in this state were right above.  so i would flip the order of ur tests above.